### PR TITLE
Create manifests for deploying the sample-apiserver

### LINKED ISSUE #25: Add manifests for deploying sample-apiserver
Adding manifests for the `sample-apiserver` would ensure users can easily learn more about the EtcdProxy controller and start using it.

This is required for E2E tests (#24).

### PR DESCRIPTION
**Pull request by xmudrii**
This PR adds manifests for deploying the [`sample-apiserver`](https://github.com/kubernetes/sample-apiserver) and updates the documentation to reflect changes made in this PR.

The manifests are located in the `deployment` directory, along with the README file containing details about the manifests and instructions how to use them.

The manifest for deploying the EtcdProxy is called `00-etcdproxy-controller.yaml` and is not dependent on any other manifests. This manifest has been added in #3.

The following manifests are used to deploy the `sample-apiserver`:
* `01-sample-apiserver-prerequisites.yaml` – creates a namespace, Service and Role/RoleBindings needed by the `sample-apiserver`.
* `02-sample-apiserver-certs.yaml` – deploys the CA and client certificate for the `sample-apiserver`, and creates ConfigMap, Secrets and RBAC roles for the proxied etcd certificates.
* `03-sample-apiserver-deployment.yaml` – creates an EtcdStorage resource, which provides the etcd storage for the API server, `sample-apiserver` ReplicaSet, and APIService to register the `sample-apiserver` with the Kubernetes API.

The ReplicaSet uses the [`xmudrii/kube-sample-apiserver`](https://hub.docker.com/r/xmudrii/kube-sample-apiserver/) image, based on Alpine 3.6.

Closes #25 
Required by #24 

/cc @deads2k @sttts

### CODE REVIEW

File: artifacts/deployment/01-apiserver-prerequisites.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  name: apiserver
+  namespace: kube-sample-apiserver
+---
+# APIService used to register the API and the API resources with the Kubernetes API.
+apiVersion: apiregistration.k8s.io/v1beta1
+kind: APIService
+metadata:
+  name: v1alpha1.wardle.k8s.io
+spec:
+  insecureSkipTLSVerify: true
```
Comment (sttts) [inline artifacts/deployment/01-apiserver-prerequisites.yaml]: TODO: make this false?

File: artifacts/deployment/01-apiserver-prerequisites.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+kind: Namespace
+metadata:
+  name: kube-sample-apiserver
+---
+# ServiceAccount to be used by the kube-sample-apiserver.
+kind: ServiceAccount
+apiVersion: v1
+metadata:
+  name: apiserver
+  namespace: kube-sample-apiserver
```
Comment (sttts) [inline artifacts/deployment/01-apiserver-prerequisites.yaml]: nit: k8s-sample-apiserver ?

File: artifacts/deployment/02-apiserver-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  template:
+    metadata:
+      labels:
+        apiserver: sample-apiserver
+    spec:
+      containers:
+      - name: sample-apiserver
+        image: xmudrii/kube-sample-apiserver
+        imagePullPolicy: Always
+        command: ["/kube-sample-apiserver", "--etcd-servers=http://wardle.kube-apiserver-storage.svc:2379"]
```
Comment (sttts) [inline artifacts/deployment/02-apiserver-deployment.yaml]: missing the configmap and secret volume. At least put them here, even though you don't use the certs yet.

File: artifacts/deployment/README.md [hunk 4]
```diff
@@ -1,26 +1,50 @@
...
-* **deployment/etcdproxy-controller-deployment** - Deployment for the `EtcdProxyController`.
 
-## RBAC Rules
+* **customresourcedefinition/etcdstorages.etcd.xmudrii.com** - CRD defining the EtcdStorage type for managing etcd proxies,
+* **deployment/etcdproxy-controller-deployment** - Controller Deployment.
+
+## Deploying the core `etcd`
+
+The deployment manifest assumes you have the core `etcd` deployed and exposed on `https://etcd-svc-1.etcd.svc:2379`.
+The URL can be changed by modifying the `--etcd-core-url` flag.
```
Comment (sttts) [inline artifacts/deployment/README.md]: ... flag in the file .....yaml

File: artifacts/deployment/README.md [hunk 5]
```diff
@@ -1,26 +1,50 @@
...
+
+The core `etcd` can be deployed using the manifests located in the [`artifacts/etcd`](https://github.com/xmudrii/etcdproxy-controller/tree/master/artifacts/etcd) directory.
```
Comment (sttts) [inline artifacts/deployment/README.md]: There is an example etcd deployment located ...

File: artifacts/deployment/README.md [hunk 6]
```diff
@@ -1,26 +1,50 @@
...
+
+The CA certificate is deployed in the ConfigMap called `etcd-coreserving-ca`, in the controller namespace. The ConfigMap name can be changed by running the controller with `--etcd-core-ca-configmap` flag.
+
+The client certificate and key are both deployed in the generic Secret called `etcd-coreserving-cert`. The Secret name can be changed by running the controller with `--etcd-core-certs-secret` flag.
+
+Deploying an EtcdStorage resoruce without the requrired ConfigMap and Secret in place causes the EtcdProxy pod to hang in the `Creating` condition.
+
+## Creating the proxied `etcd`
+
+To create the etcd instance for the aggregated API server, deploy the `EtcdStorage` manifest. Once the manifest is deployed, the controller creates and exposes the etcd proxy running against the `etcd` namespace named same as the `EtcdStorage` resource.
```
Comment (sttts) [inline artifacts/deployment/README.md]: EtcdProxy instance

File: artifacts/deployment/01-apiserver-prerequisites.yaml (hunk 2 above)
Comment (xmudrii) [reply artifacts/deployment/01-apiserver-prerequisites.yaml]: Sounds good to me. Updated

File: artifacts/deployment/README.md (hunk 4 above)
Comment (xmudrii) [reply artifacts/deployment/README.md]: Updated. Fixed several other lines as well.

File: artifacts/deployment/README.md (hunk 5 above)
Comment (xmudrii) [reply artifacts/deployment/README.md]: Done.

File: artifacts/deployment/README.md (hunk 6 above)
Comment (xmudrii) [reply artifacts/deployment/README.md]: Done.

File: artifacts/deployment/02-apiserver-deployment.yaml (hunk 3 above)
Comment (xmudrii) [reply artifacts/deployment/02-apiserver-deployment.yaml]: I have created manifests for deploying the Secret, ConfigMap, and RBAC roles to ensure needed permissions for the controller SA. Also, updated the Deployment to use them.

File: artifacts/deployment/03-sample-apiserver-deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        secret:
+          secretName: etcd-client-cert
+---
+# APIService used to register the API and the API resources with the Kubernetes API.
+apiVersion: apiregistration.k8s.io/v1beta1
+kind: APIService
+metadata:
+  name: v1alpha1.wardle.k8s.io
+spec:
+  insecureSkipTLSVerify: true
```
Comment (xmudrii) [inline artifacts/deployment/03-sample-apiserver-deployment.yaml]: Add TODO about certs or add them in this PR.

**Comment by deads2k** [comment]
> My question here is, is this acceptable, or should I try to find a solution letting us to enable the TLS checks?

I don't care about TLS checks for the aggregator to apiservice connection at the moment.

**Comment by xmudrii** [comment]
@deads2k Sounds good to me. In that case, I think we have the complete `sample-apiserver` manifests. I've updated the PR to fix some typos and a mistake made in the command for `sample-apiserver`. Also, I've tested the manifests once again to make sure they're working as intended.

File: artifacts/deployment/00-etcdproxy-controller.yaml [hunk 8]
```diff
@@ -127,4 +127,4 @@ spec:
         command:
           - /etcdproxy-controller
           - "--etcd-core-url=https://etcd-svc-1.etcd.svc:2379"
-        imagePullPolicy: Always
+        imagePullPolicy: Always
```
Comment (deads2k) [inline artifacts/deployment/00-etcdproxy-controller.yaml]: nit, i personally prefer the newline at the end.

File: artifacts/deployment/01-sample-apiserver-prerequisites.yaml [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: extension-apiserver-authentication-reader
+subjects:
+- kind: ServiceAccount
+  name: apiserver
+  namespace: k8s-sample-apiserver
+---
+# ClusterRole used to allow API server to use Admission Webhooks.
+kind: ClusterRole
```
Comment (deads2k) [inline artifacts/deployment/01-sample-apiserver-prerequisites.yaml]: We should actually be providing this in core kube.  Can you open an issue and lay out the case for an aggregated cluster role for aggregated API servers?

File: artifacts/deployment/00-etcdproxy-controller.yaml (hunk 8 above)
Comment (xmudrii) [reply artifacts/deployment/00-etcdproxy-controller.yaml]: Fixed.

File: artifacts/deployment/01-sample-apiserver-prerequisites.yaml (hunk 9 above)
Comment (xmudrii) [reply artifacts/deployment/01-sample-apiserver-prerequisites.yaml]: An issue in the `kubernetes/kubernetes` repo?

File: artifacts/deployment/02-sample-apiserver-certs.yaml [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+# are stored in the 'sample-apiserver-clientcert' Secret.
+#
+# The CA certificate for the proxied etcd is stored in the 'etcd-serving-ca' ConfigMap,
+# and the client certificate and key are stored in the 'etcd-client-cert' Secret.
+#
+# The etcd certificates are managed by the EtcdProxy Controller, so this manifests
+# creates RBAC roles allowing the EtcdProxy Controller ServiceAccount to get, update and patch
+# ConfigMap and Secret containing the etcd certificates.
+---
+# The CA certificate for the sample-apiserver. Allows users to use client certificates to access the API.
```
Comment (deads2k) [inline artifacts/deployment/02-sample-apiserver-certs.yaml]: we can't read the one from kube-public or kube-system?  You can have it for now, but it's weird.

File: artifacts/deployment/02-sample-apiserver-certs.yaml [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+---
+# ConfigMap used to store the CA certificate for the proxied etcd.
```
Comment (deads2k) [inline artifacts/deployment/02-sample-apiserver-certs.yaml]: to verify the etcd proxy serving certificate?

File: artifacts/deployment/03-sample-apiserver-deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+---
+# Deploys the sample-apiserver (https://github.com/kubernetes/sample-apiserver).
+# The Deployment uses the xmudrii/kube-sample-apiserver image.
+apiVersion: apps/v1
+kind: ReplicaSet
+metadata:
+  name: apiserver
+  namespace: k8s-sample-apiserver
+spec:
+  replicas: 2
```
Comment (deads2k) [inline artifacts/deployment/03-sample-apiserver-deployment.yaml]: just 1 I'd think.   It's a demo.

File: artifacts/deployment/02-sample-apiserver-certs.yaml (hunk 10 above)
Comment (xmudrii) [reply artifacts/deployment/02-sample-apiserver-certs.yaml]: Yes, we can, and it's the default behavior. What I wanted is to enable users to use API outside, but, I can agree it's bit weird.

Do you want me to get rid of those?

**Comment by deads2k** [comment]
You'll add the etcdstorage manifest later?

I have minor nits.  Feel free to self-merge after addressing

File: artifacts/deployment/02-sample-apiserver-certs.yaml (hunk 10 above)
Comment (deads2k) [reply artifacts/deployment/02-sample-apiserver-certs.yaml]: > Yes, we can, and it's the default behavior. What I wanted is to enable users to use API outside, but, I can agree it's bit weird.

Assuming it wires well by default, you can use your kube client cert to speak directly to the sample apiserver.

Comment (xmudrii) [reply artifacts/deployment/02-sample-apiserver-certs.yaml]: Removed them for now. It should be the default behavior

**Pull request merged**